plugins/javaembed/HpccClassLoader.java

```diff
@@ -21,47 +21,60 @@ HPCC SYSTEMS software Copyright (C) 2018 HPCC Systems(R).
 {
     private long bytecode;
     private int bytecodeLen;
+    private Boolean trace = false;
     private native Class<?> defineClassForEmbed(int bytecodeLen, long bytecode, String name);
     private Hashtable<String, Class<?>> classes = new Hashtable<>();
-    static private Hashtable<String, java.net.URLClassLoader> pathLoaders = new Hashtable<>();
-    private java.net.URLClassLoader pathLoader;
+    static private Hashtable<String, java.net.URLClassLoader> sharedPathLoaders = new Hashtable<>();
+    private List<java.net.URLClassLoader> libraryPathLoaders = new ArrayList<>();
     private HpccClassLoader(String classPath, ClassLoader parent, int _bytecodeLen, long _bytecode, String dllname)
     {
         super(parent);
         if (classPath != null && !classPath.isEmpty())
         {
-            synchronized(pathLoaders)
+            String[] libraryPaths = classPath.split("\\|");  // Careful - the param is a regex so need to escape the |
+            synchronized(sharedPathLoaders)
             {
-                pathLoader = pathLoaders.get(classPath);
-                if (pathLoader == null)
+                for (String libraryPath : libraryPaths)
                 {
-                    List<URL> urls = new ArrayList<>();
-                    String[] paths = classPath.split(";");
-                    for (String path : paths)
+                    URLClassLoader libraryPathLoader = sharedPathLoaders.get(libraryPath);
+                    if (libraryPathLoader == null)
                     {
-                        try
-                        {
-                            if (path.contains(":"))
-                                urls.add(new URL(path));
-                            else
-                                urls.add(new URL("file:" + path));
-                        }
-                        catch (MalformedURLException E)
+                        List<URL> urls = new ArrayList<>();
+                        String[] paths = libraryPath.split(";");
+                        for (String path : paths)
                         {
-                            // Ignore any that we don't recognize
-                            // System.out.print(E.toString());
+                            try
+                            {
+                                if (path.contains(":"))
+                                    urls.add(new URL(path));
+                                else
+                                    urls.add(new URL("file:" + path));
+                            }
+                            catch (MalformedURLException E)
+                            {
+                                // Ignore any that we don't recognize
+                                if (trace)
+                                    HpccUtils.log("Malformed URL: " + E.toString());
+                            }
                         }
+                        libraryPathLoader = new URLClassLoader(urls.toArray(new URL[urls.size()]));
```

Comment:
I think it would be better to pass `parent` into the URLClassLoader as it's parent.  As it's written here the URLClassLoader parent will be the default class loader for the thread.  I expect that's usually the same as the `parent` parameter here but it would be safer to pass it in to make sure the same parent is used everywhere here.

Reply:
Actually I suspect I should be passing null for the parent. Lookup using the default class loader should happen after ALL the URLClassLoaders have had a bite...